android/KMEA/app/src/main/java/com/keyman/engine/KMKeyboardJSHandler.java

```diff
@@ -205,7 +205,63 @@ public void setIsChiral(boolean isChiral) {
   }
 
   @JavascriptInterface
-  public abstract boolean dispatchKey(final int code, final int eventModifiers);
+  public boolean dispatchKey(final int code, final int eventModifiers) {
+    Handler mainLoop = new Handler(Looper.getMainLooper());
+    mainLoop.post(new Runnable() {
+      public void run() {
+        if (k == null) {
+          KMLog.LogError(TAG, "dispatchKey failed: Keyboard is null");
+          return;
+        }
+
+        if (k.subKeysWindow != null) {
+          return;
+        }
+
+        if (k.keyboardType == KeyboardType.KEYBOARD_TYPE_INAPP &&
+            (KMTextView.activeView == null || KMTextView.activeView.getClass() != KMTextView.class)) {
+          if (KMTextView.activeView == null && KMManager.isDebugMode()) {
+            Log.w(TAG, "dispatchKey failed: activeView is null");
+          }
+          return;
+        }
```

Comment:
Can we split this into separate tests to make it clearer what is happening and that one of these conditions is an error case? Also, are similar checks going to be repeated elsewhere?

```suggestion
        if (k.keyboardType == KeyboardType.KEYBOARD_TYPE_INAPP) {
          if (KMTextView.activeView == null) {
            if (KMManager.isDebugMode()) {
              Log.w(TAG, "dispatchKey failed: activeView is null");
            }
            return;
          }
          if (KMTextView.activeView.getClass() != KMTextView.class)) {
            return;
          }
        }
```

Separate question on `isDebugMode`: is there any need to predicate on that for emitting warnings to the log? The log is not generally visible to end users anyway, and it simplifies our code significantly to remove all those `if (KMManager.isDebugMode())` checks!

Reply:
I had to dig through the PR chronicles to #837 where most of the debug logging got removed. I think we maybe we convert these to Sentry logs (shouldn't be happening?)

Reply:
I originally accepted the suggestion, but later turned into a helper `inappKMTextViewInvalid()` to use within `insertText` and `dispatchKey`.